lazy-pages/src/sys/unix.rs

```diff
@@ -109,7 +109,102 @@ where
     }
 }
 
-pub unsafe fn setup_signal_handler<H>() -> io::Result<()>
+use errno::Errno;
+
+#[derive(Debug, Clone, Copy, derive_more::Display)]
+enum ThreadInitError {
+    #[display(fmt = "Cannot get information about old signal stack: {}", _0)]
+    OldStack(Errno),
+    #[display(fmt = "Cannot mmap space for signal stack: {}", _0)]
+    Mmap(Errno),
+    #[display(fmt = "Cannot set new signal stack: {}", _0)]
+    SigAltStack(Errno),
+}
+
+fn init_for_thread_internal() -> Result<(), ThreadInitError> {
+    use core::{mem, ptr};
+
+    // Should be enough for lazy-pages signal handler.
+    // Equal to libc::SIGSTKSZ on macos M1.
+    const SIGNAL_STACK_SIZE: usize = 0x20000;
+
+    enum StackInfo {
+        UseOldStack,
+        NewStack(*mut libc::c_void),
+    }
+
+    impl Drop for StackInfo {
+        fn drop(&mut self) {
+            if let StackInfo::NewStack(mmap_ptr) = self {
+                unsafe {
+                    // Deallocate the stack memory.
+                    if libc::munmap(*mmap_ptr, SIGNAL_STACK_SIZE) != 0 {
+                        log::error!(
+                            "Cannot deallocate signal stack memory during the thread shutdown: {}",
+                            errno::errno()
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    unsafe fn init_sigstack() -> Result<StackInfo, ThreadInitError> {
+        // Check whether old signal stack exist and suitable for lazy-pages signal handler.
+        let mut old_stack = mem::zeroed();
+        let res = libc::sigaltstack(ptr::null(), &mut old_stack);
+        if res != 0 {
+            return Err(ThreadInitError::OldStack(errno::errno()));
+        }
+        if old_stack.ss_flags & libc::SS_DISABLE == 0 && old_stack.ss_size >= SIGNAL_STACK_SIZE {
+            return Ok(StackInfo::UseOldStack);
+        }
+
+        // Alloc memory for new signal stack.
+        let ptr = libc::mmap(
```

Comment:
Add comment why you use mmap, if you motivation is taken here: https://stackoverflow.com/a/31820184/7844786

Reply:
I use here mmap because it's more clear what it does in this place. I don't wanna use malloc or region::alloc because they can have side effects which I must investigate before using. Using mmap here is the most clear variant

Reply:
Make comment